rkyv/src/niche/niching.rs

```diff
@@ -2,75 +2,132 @@
-/// struct One;
+/// // Let's niche `Option<u32>` by using odd values
+/// struct NeverOdd;
+///
+/// unsafe impl Niching<ArchivedU32> for NeverOdd {
+///     type Niched = ArchivedU32;
 ///
-/// // SAFETY: `Self::Niched` is the same as `T` so it's always valid to access
-/// // it within a union of the two. Furthermore, we can be sure that the `T`
-/// // field is safe to access if `is_niched` returns `false`.
-/// unsafe impl Niching<Archived<u32>> for One {
-///     type Niched = Archived<u32>;
+///     unsafe fn niched_ptr(
+///         ptr: *const ArchivedU32,
+///     ) -> Option<*const Self::Niched> {
+///         // We niche into the same type, no casting required
+///         Some(ptr)
+///     }
 ///
-///     fn is_niched(niched: &Self::Niched) -> bool {
-///         *niched == 1
+///     unsafe fn is_niched(niched: *const ArchivedU32) -> bool {
+///         // Interprete odd values as "niched"
+///         unsafe { *niched % 2 == 1 }
 ///     }
 ///
-///     fn resolve_niched(out: Place<Self::Niched>) {
-///         u32::resolve(&1, (), out);
+///     fn resolve_niched(out: Place<ArchivedU32>) {
+///         // To niche, we use the value `1`
+///         out.write(ArchivedU32::from_native(1))
 ///     }
 /// }
 ///
 /// #[derive(Archive)]
-/// struct Basic(Option<u32>);
+/// struct Basic {
+///     field: Option<u32>,
+/// }
 ///
 /// #[derive(Archive, Serialize)]
-/// struct Niched(#[rkyv(with = Nicher<One>)] Option<u32>);
+/// struct Niched {
+///     #[rkyv(with = Nicher<NeverOdd>)]
+///     field: Option<u32>,
+/// }
 ///
 /// # fn main() -> Result<(), rkyv::rancor::Error> {
+/// // Indeed, we have a smaller archived representation
 /// assert!(size_of::<ArchivedNiched>() < size_of::<ArchivedBasic>());
 ///
-/// let values = [Niched(Some(1)), Niched(Some(42)), Niched(None)];
+/// let values: Vec<Niched> =
+///     (0..4).map(|n| Niched { field: Some(n) }).collect();
+///
 /// let bytes = rkyv::to_bytes(&values)?;
-/// let archived = rkyv::access::<[ArchivedNiched; 3], _>(&bytes)?;
-/// assert_eq!(archived[0].0.as_ref(), None);
-/// assert_eq!(archived[1].0.as_ref(), Some(&42.into()));
-/// assert_eq!(archived[2].0.as_ref(), None);
+/// let archived = rkyv::access::<Archived<Vec<Niched>>, _>(&bytes)?;
+/// assert_eq!(archived[0].field.as_ref(), Some(&0.into()));
+/// assert_eq!(archived[1].field.as_ref(), None);
+/// assert_eq!(archived[2].field.as_ref(), Some(&2.into()));
+/// assert_eq!(archived[3].field.as_ref(), None);
 /// # Ok(()) }
 /// ```
 ///
+/// [`MaybeUninit<T>`]: core::mem::MaybeUninit
 /// [`Nicher`]: crate::with::Nicher
+/// [`Portable`]: crate::traits::Portable
 /// [`is_niched`]: Niching::is_niched
+/// [`resolve_niched`]: Niching::resolve_niched
+/// [`niched_ptr`]: Niching::niched_ptr
 pub unsafe trait Niching<T> {
-    /// The archived representation of a niched value.
-    type Niched: Portable;
+    /// The type that is leveraged for niching.
+    type Niched;
```

Comment:
Pretty much the only reason for the associated type's existence is the `CheckBytes` implementation of `NichedOption`. There are basically two ways to go about it:
- Have a method in the `Niching` trait to check the niched value. This seems unfavorable since checking types is behind a feature flag and modifying traits based on that isn't best practice.
- Let `Niching` describe what type will be used for niching and how to access it so that callers can use that when necessary. Despite the added noise in the trait, this seemed like the better approach to me.

At least those were the only options I could come up with to implement `CheckBytes` for `NichedOption`.

Reply:
I think we can say that "if is_niched() returns false, there is no valid value in the MaybeUninit and it doesn't need to be checked". Would that clear up the issue?